File: textEditor.pl
#!/usr/bin/perl
# Hannah Gulle
# Text editor to compute the Flesch Readability Index 
# and the Flesch Kincaid Grade Level Index using Perl.
# August 30, 2017

# Opening File Now

my $infile = <infile>;
open($infile, '<:encoding(UTF-8)', $infile) or die "Yo No File Son";

read $infile, $file_string, -s $infile;		# -s is length of file
close $infile;

$syllables = 0.0;
$words = 0.0;
$sentences = 0.0;
$syllFound = 0;

foreach $curr (split //, $file_string){
	if(not isDigit($curr)){
		if( (not isVowel($prev))  & isVowel($curr)){
			$syllables++;
			$syllFound = 1;		
		}
		elsif( isPunct($curr) ){
			$sentences++;
			$words++;
			if($syllFound == 0){
				$syllables++;
			}
		}
		elsif( $prev eq 'e'){
			if( (isPunct($curr)) || isDelim($curr)){
				$syllables--;
			}
		}
		elsif( isAlpha($prev) & isDelim($curr)){
			$words++;
			if($syllFound == 0){
				$syllables++;
			}
		}
		$prev = $curr;
	}
}

print "Syllables: $syllables \n";
print "Words: $words \n";
print "Sentences: $sentences \n";

$alpha = ($syllables/$words);
$beta = ($words/$sentences);
$flesch = flesch($syllables, $words, $sentences);
$fkincaid = fkincaid($syllables, $words, $sentences);

my $printFlesch = sprintf("%.0f", $flesch);
my $printFleschKincaid = sprintf("%0.1f", $fkincaid);

print "flesch: $printFlesch \n";
print "fkincaid: $printFleschKincaid \n";

sub flesch{
	my($syllables, $words, $sentences) = @_;
	$alpha = ($syllables/$words);
	$beta = ($words/$sentences);
	return(206.835 - ($alpha*84.6) - ($beta*1.015));
}

sub fkincaid{
	my($syllables, $words, $sentences) = @_;
	$alpha = ($syllables/$words);
	$beta = ($words/$sentences);
	return((11.8*$alpha) + ($beta*.39) - 15.59);
}

sub isVowel{
	my($curr) = @_;
	$vowels = "aeiouy";
	return(index($vowels, $curr) != -1);
}

sub isDelim{
	my($curr) = @_;
	$delims = " ,]";
	return(index($delims, $curr) != -1);
}

sub isPunct{
	my($curr) = @_;
	$punct = ".:;?!";
	return(index($punct, $curr) != -1);
}

sub isDigit{
	my($curr) = @_;
	$digits = "0123456789";
	return(index($digits, $curr) != -1);
}

sub isAlpha{
	my($curr) = @_;
	$alpha = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
	return(index($alpha, $curr) != -1);
}

File: textEditor.f90
! Hannah Gulle
! September 28, 2017
! Text parsing program to compute the Flesch Readability Index and the
! Flesch Kincaid Grade Level index using Fortran 90

program texteditor

integer::filesize

! -----------------------------
! read_file interface
! ----------------------------
interface
subroutine read_file(filesize)
integer::filesize
end subroutine read_file
end interface


call read_file(filesize)
end program texteditor

! -------------------------------------------------------------------------------
! read_file subroutine
! ------------------------------------------------------------------------------

subroutine read_file(filesize)
logical::syllFound
integer::syllables,words,sentences
character(LEN=1)::input
character(LEN=1)::curr,prev
integer::counter, filesize
real*8::fkincaid
integer::flesch
real*4::aVal,bVal
syllables = 0
words = 0
sentences = 0


! get number of characters in file
open(unit=5,status="old",access="direct",form="unformatted",recl=1,file="KJV.txt")

! put file in string of correct number of characters
counter = 1
prev = '~'
syllFound = .false.
100 read(5,rec=counter,err=200)input
        curr=input

        if( isDigit(curr) == 0) then
                if( (isVowel(prev) == 0) .and. (isVowel(curr) /= 0)) then
                        syllables = syllables+1
                        syllFound = .true.
                elseif( isPunct(curr) /= 0) then
                        sentences = sentences +1
                        words = words+1
                        if( syllFound .eqv. .false.) then
                                syllables = syllables+1
                        endif
                elseif( (index("e",prev) /= 0) .and. (index(" ,].:;?!",curr) /= 0)) then
                                syllables = syllables-1
                elseif( isAlpha(prev) /= 0 .and. (index(" ,]",curr) /= 0)) then
                        words = words +1
                        if(syllFound .eqv. .false.)then
                                syllables = syllables+1
                        endif
                endif
                prev = curr
        endif
        counter = counter + 1
        goto 100
200 continue
close(5)

print*, "Syllables ",syllables
print*, "Words ", words
print*, "sentences ", sentences

aVal = real(syllables)/real(words)
bVal = real(words)/real(sentences)

flesch = computeFlesch( aVal, bVal)
fkincaid = computefKincaid( aVal, bVal )

write(*,20) "Flesch: ",flesch
20 format(a,i3)

write(*,30) "FleschKincaid: ",fkincaid
30 format(a,f3.1)

end subroutine read_file

function computeFlesch(alpha, beta)
real*4::alpha,beta
computeFlesch = ((206.835 - (84.6*alpha)) - (1.015*beta));
return
end function computeFlesch

function computefKincaid(alpha,beta)
real*4::alpha,beta
computefKincaid = ( ((11.8*alpha) + (0.39*beta)) - 15.59)
return
end function computefKincaid

function isVowel(curr)
character(LEN=1)::curr
character(LEN=6)::vowels
vowels = "aeiouy"
isVowel = ( index(vowels, curr) )
return
end function isVowel

function isPunct(curr)
character(LEN=1)::curr
character(LEN=5)::punct
punct = ".:;?!"
isPunct = (index(punct, curr) )
return
end function isPunct

function isDigit(curr)
character(LEN=1)::curr
character(LEN=10)::digit
digit = "0123456789"
isDigit = (index(digit,curr) )
return 
end function isDigit

function isAlpha(curr)
character(LEN=1)::curr
character(LEN=52)::alpha
alpha = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
isAlpha = (index(alpha,curr) )
return
end function isAlpha
